src/versioning.rs

```diff
@@ -0,0 +1,540 @@
+                VersioningState::ValidDowngrade => {
+                    let message = format!(
+                        "Downgrading from [{}] to [{}]",
+                        current_version, spec_version
+                    );
+
+                    info!("{}", message);
+
+                    let condition = build_versioning_condition(
+                        resource,
+                        conditions,
+                        &message,
+                        VersioningConditionReason::Downgrading.as_ref(),
+                        ConditionStatus::True,
+                    );
+
+                    let version = build_version(None, Some(spec_version));
+
+                    (Some(version), Some(condition))
+                }
+                VersioningState::NoOp => {
+                    let message = format!(
+                        "No upgrade required [{}] is still the current_version",
+                        current_version
+                    );
+
+                    debug!("{}", message);
+
+                    (None, None)
+                }
+                VersioningState::NotSupported => {
+                    warn!("Up-/Downgrade from [{}] to [{}] not possible but requested in spec: Ignoring, will continue \
+                           reconcile as if the invalid version weren't set", current_version, spec_version);
+                    (None, None)
+                }
+                VersioningState::Invalid(err) => {
+                    // TODO: throw error
+                    error!("Error occurred for versioning: {}", err);
+                    (None, None)
+                }
+            }
+        }
+        _ => (None, None),
+    };
+}
+
+/// Builds a condition for versioning. It basically forwards every parameter and uses the
+/// fixed `CONDITION_TYPE` to identify the versioning condition.
+///
+/// # Arguments
+///
+/// * `resource` - The cluster custom resource.
+/// * `conditions` - The conditions from the custom resource status.
+/// * `message` - The message set in the conditions.
+/// * `reason` - The reason set in the conditions
+/// * `status` - The status set in the conditions.
+///
+fn build_versioning_condition<T>(
+    resource: &T,
+    conditions: &[Condition],
+    message: &str,
+    reason: &str,
+    status: ConditionStatus,
+) -> Condition
+where
+    T: Clone + Debug + DeserializeOwned + Resource<DynamicType = ()>,
+{
+    build_condition(
+        resource,
+        Some(conditions),
+        message.to_string(),
+        reason.to_string(),
+        status,
+        CONDITION_TYPE.to_string(),
+    )
+}
+
+/// Builds a `ProductVersion` to be written into the custom resource status.
+///
+/// # Arguments
+///
+/// * `current_version` - The optional current version of the cluster.
+/// * `target_version` - The optional target version for upgrading / downgrading the cluster.
+///
+fn build_version<V>(current_version: Option<V>, target_version: Option<V>) -> ProductVersion<V> {
```

Comment:
I think `ProductVersion::new()` is more conventional.

Reply:
yeah that is definitly right, need to adapt :D